### PR TITLE
chore: initial bootstrap

### PR DESCRIPTION
**Pull request by katcipis**
Some initial automation/CI/structure. The idea is to have two crates, all logic on the library crate and a very small binary crate that mostly calls the library crate (and some flag parsing at some point). This should make even end to end tests easier/possible. Always loved doing that in C and I find it cool that it is considered the [best way to structure binaries in Rust](https://doc.rust-lang.org/book/ch12-03-improving-error-handling-and-modularity.html#separation-of-concerns-for-binary-projects) :-).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,50 @@
...
+have a secure small network and want to chat with people inside it.
+
+Not a novel idea... We know... but we also want to play around with some Rust :-).
+
+The major architectural constraints are:
+
+1 - Communication is always peer to peer
+2 - Any storage (if any) is stored only on the peers participating on the conversation
+3 - The network is secure (protocols will be plain text/binary, no TLS)
+4 - Peer to peer communication is always possible
```
Comment (lfritz) [inline README.md]:
```suggestion
 1. Communication is always peer to peer
 2. Any storage (if any) is stored only on the peers participating on the conversation
 3. The network is secure (protocols will be plain text/binary, no TLS)
 4. Peer to peer communication is always possible
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,50 @@
...
+
+The constraints and the overall design are inspired on the ideas of [Remembering the LAN](https://crawshaw.io/blog/remembering-the-lan),
```
Comment (lfritz) [inline README.md]:
```suggestion
The constraints and the overall design are inspired on the ideas by [Remembering the LAN](https://crawshaw.io/blog/remembering-the-lan),
```

File: README.md [hunk 3]
```diff
@@ -1,2 +1,50 @@
...
+The LAN was a magical place to learn about computers.
+Besides the physical aspect of assembling and disassembling machines,
+I could safely do things unthinkable on the modern internet: permission-less file sharing,
+experimental servers with no security, shared software where any one machine could easily bring down
+the network by typing in an innocuous command.
+Even when I did bring down the network the impact never left the building.
+I knew who I had to apologise to.
+
+With our LAN easy things were easy, and some hard things were possible.
+```
```
Comment (lfritz) [inline README.md:30]:
```suggestion
+> The LAN was a magical place to learn about computers.
+> Besides the physical aspect of assembling and disassembling machines,
+> I could safely do things unthinkable on the modern internet: permission-less file sharing,
+> experimental servers with no security, shared software where any one machine could easily bring down
+> the network by typing in an innocuous command.
+> Even when I did bring down the network the impact never left the building.
+> I knew who I had to apologise to.
+>
+> With our LAN easy things were easy, and some hard things were possible.
```

File: README.md [hunk 4]
```diff
@@ -1,2 +1,50 @@
...
+
+Technology like wireguard/tailscale helps us go back to those simpler times by building "LANs" on top
+of the Internet. We want to explore building things on top of this model, where application code
+can just trust other peers and doesn't need to deal with the hassle of making peer to peer communication
```
Comment (lfritz) [inline README.md]:
```suggestion
can just trust other peers and doesn't need to deal with the hassle of making peer-to-peer communication
```

**Review by lfritz — APPROVED** [review]
Some small suggestions for the README

File: README.md (hunk 3 above)
Comment (katcipis) [reply README.md:30]: > 

It is so annoying/a hassle to add > on the beginning of every single line...I kinda like the block behavior =P. The end result is much better ?

Comment (lfritz) [reply README.md:30]: Yeah, if you prefer the ‘’’ formatting, ignore this suggestion :)

**Pull request merged**